### ISSUE TITLE
Allow Using Specific Dataloaders in LightningCLI (Instead of a DataModule)

### ISSUE THREAD

**Issue by stuartthomson** [Lightning-AI/pytorch-lightning#20673]
### Description & Motivation

I would like to configure the LightningCLI to use dataloaders in the `test` subcommand, similar to this example from [the docs](https://lightning.ai/docs/pytorch/stable/common/trainer.html#testing)
```python
trainer.test(dataloaders=test_dataloaders)
```

As far as I can tell this impossible using the LightningCLI because you can only pass a `datamodule` into it. My test set is defined differently from various training sets I have so this becomes annoying to create separate datamodules for each.

You can see that the `dataloaders` option is specifically ignored in [the `cli.py` `subcommands` method](https://github.com/Lightning-AI/pytorch-lightning/blob/df5dee674243e124a2bf34d9975dd586ff008d4b/src/lightning/pytorch/cli.py#L476).

### Pitch

I would like to specify `dataloaders` in the `LightningCLI`. Perhaps like:

```python
LightningCLI(
  MyModule,
  train_dataloaders=MyTrainDataloader,
  test_dataloaders=MyTestDataloader,
  args=args
)
```

### Alternatives

The alternative is having to create different datamodules for every combination of train/test dataloaders.
Or you have to simply just not use the CLI stuff and call lightning directly.

**Comment by mauvilsa** [comment]
I don't know why [these](https://github.com/Lightning-AI/pytorch-lightning/blob/2c74bee068a664e124cc025155ba4bd563b76630/src/lightning/pytorch/cli.py#L493-L496) arguments are skipped. But note that `LightningCLI.subcommands` is public. This means you can subclass `LightningCLI` and modify `subcommands` such that `dataloaders` doesn't get skipped. It is your choice. There is no really a need to fix lightning code. Maybe what could be done is to mention this in the docs in [lightning_cli_expert.rst](https://github.com/Lightning-AI/pytorch-lightning/blob/master/docs/source-pytorch/cli/lightning_cli_expert.rst).

**Comment by stuartthomson** [comment]
Your suggestion is good and will help unblock this specific case but I'm unsure why it should be necessary

**Comment by mauvilsa** [comment]
I don't know either. I think this was implemented by @carmocca. Any insight here?

**Comment by carmocca** [comment]
Sorry, I don't remember exactly. But I guess it was to keep the design simpler. The LightningCLI accepts a model and a datamodule, each of which can already define the dataloaders. So there's ambiguity about what should be called if these extra arguments are present. There's probably not a technical limitation to support it.

**Comment by mauvilsa** [comment]
Thank you for the explanation. Even if the ambiguity is not important, I am not sure about changing it. Since `def subcommands()` is public, changes could break other people's code.
